File: app/src/main/java/com/minerdev/greformanager/HouseDetailActivity.java
package com.minerdev.greformanager;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;
import android.widget.ToggleButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.BindingConversion;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import com.android.volley.Request;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.minerdev.greformanager.databinding.ActivityHouseDetailBinding;
import com.naver.maps.geometry.LatLng;
import com.naver.maps.map.CameraUpdate;
import com.naver.maps.map.MapFragment;
import com.naver.maps.map.UiSettings;
import com.naver.maps.map.overlay.Marker;

public class HouseDetailActivity extends AppCompatActivity {
    private final ImageAdapter adapter = new ImageAdapter();
    private HouseWrapper houseWrapper;
    private int originalState;

    private ActivityHouseDetailBinding binding;
    private ImageViewModel imageViewModel;

    @BindingConversion
    public static int convertBooleanToVisibility(boolean visible) {
        return visible ? View.VISIBLE : View.GONE;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_house_detail);
        binding.setActivity(this);

        imageViewModel = new ViewModelProvider(this,
                new ViewModelProvider.AndroidViewModelFactory(getApplication())).get(ImageViewModel.class);


        Intent intent = getIntent();
        House data = intent.getParcelableExtra("house_value");
        originalState = data.state;
        houseWrapper = new HouseWrapper(data);


        setSupportActionBar(binding.houseDetailToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        imageViewModel.getOrderByPosition(data.id).observe(this, images -> {
            Log.d("DB_CHANGED", "observe");

            for (Image image : images) {
                Log.d("DB_CHANGED", image.title);
            }

            adapter.addImages(images);
            adapter.notifyDataSetChanged();
        });

        binding.houseDetailViewPager2Image.setAdapter(adapter);


        setMapFragment();


        ToggleButtonGroup toggleButtonGroupManageFee = new ToggleButtonGroup(this, "ManageFee");
        toggleButtonGroupManageFee.addToggleButtonsFromText(houseWrapper.getManageFeeContains());
        for (ToggleButton toggleButton : toggleButtonGroupManageFee.getToggleButtons()) {
            binding.houseDetailFlowLayoutManageFee.addView(toggleButton);
        }

        ToggleButtonGroup toggleButtonGroupOptions = new ToggleButtonGroup(this, "Options");
        toggleButtonGroupOptions.addToggleButtonsFromText(houseWrapper.getOptions());
        for (ToggleButton toggleButton : toggleButtonGroupOptions.getToggleButtons()) {
            binding.houseDetailFlowLayoutOptions.addView(toggleButton);
        }


        readItems();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.house_detail_toolbar, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;

            case R.id.house_detail_menu_modify:
                Intent intent = new Intent(this, HouseModifyActivity.class);
                intent.putExtra("mode", "modify");
                intent.putExtra("house_value", houseWrapper.getData());
                startActivityForResult(intent, Constants.getInstance().HOUSE_MODIFY_ACTIVITY_REQUEST_CODE);
                break;

            default:
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void finish() {
        if (originalState != houseWrapper.getData().state) {
            JsonObject data = new JsonObject();
            data.addProperty("state", houseWrapper.getData().state);
            HttpConnection.getInstance().send(this, Request.Method.PATCH,
                    "houses/" + houseWrapper.getData().id, data, null);
        }

        Intent intent = new Intent();
        intent.putExtra("house_value", houseWrapper.getData());
        setResult(RESULT_OK, intent);

        super.finish();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == Constants.getInstance().HOUSE_MODIFY_ACTIVITY_REQUEST_CODE) {
            if (resultCode == RESULT_OK) {
                if (data == null) {
                    return;
                }

                HouseWrapper houseWrapper = new HouseWrapper(data.getParcelableExtra("house_value"));
                this.houseWrapper = houseWrapper;
                refreshUI();
            }
        }
    }

    public HouseWrapper getHouseWrapper() {
        return houseWrapper;
    }

    private void refreshUI() {
        binding.invalidateAll();
        binding.houseDetailFlowLayoutManageFee.removeAllViews();
        binding.houseDetailFlowLayoutOptions.removeAllViews();

        ToggleButtonGroup toggleButtonGroupManageFee = new ToggleButtonGroup(this, "ManageFee");
        toggleButtonGroupManageFee.addToggleButtons(houseWrapper.getManageFeeContains().split("\\|"));
        for (ToggleButton toggleButton : toggleButtonGroupManageFee.getToggleButtons()) {
            binding.houseDetailFlowLayoutManageFee.addView(toggleButton);
        }

        ToggleButtonGroup toggleButtonGroupOptions = new ToggleButtonGroup(this, "Options");
        toggleButtonGroupOptions.addToggleButtons(houseWrapper.getOptions().split("\\|"));
        for (ToggleButton toggleButton : toggleButtonGroupOptions.getToggleButtons()) {
            binding.houseDetailFlowLayoutOptions.addView(toggleButton);
        }

        setMapFragment();
    }

    private void setMapFragment() {
        MapFragment mapFragment = (MapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
        if (mapFragment == null) {
            mapFragment = MapFragment.newInstance();
            getSupportFragmentManager().beginTransaction().add(R.id.map, mapFragment).commit();
        }

        mapFragment.getMapAsync(naverMap -> {
            naverMap.setLiteModeEnabled(true);

            UiSettings uiSettings = naverMap.getUiSettings();
            uiSettings.setZoomControlEnabled(false);
            uiSettings.setAllGesturesEnabled(false);

            Geocode.getInstance().getQueryResponseFromNaver(HouseDetailActivity.this, houseWrapper.getAddress().substring(8));
            Geocode.getInstance().setOnDataReceiveListener(result -> {
                GeocodeResult.Address address = result.addresses.get(0);
                LatLng latLng = new LatLng(Double.parseDouble(address.y), Double.parseDouble(address.x));
                naverMap.moveCamera(CameraUpdate.scrollAndZoomTo(latLng, 16));
                Marker marker = new Marker(latLng);
                marker.setMap(naverMap);
            });
        });
    }

    private void readItems() {
        HttpConnection.getInstance().receive(this, "houses/" + houseWrapper.getId() + "/images",
                receivedData -> {
                    Gson gson = new Gson();
                    Image[] array = gson.fromJson(receivedData, Image[].class);
                    if (array == null) {
                        Toast.makeText(this, "데이터 수신 실패.", Toast.LENGTH_SHORT).show();
                        return;
                    }

                    Log.d("DB_DATA", receivedData);

                    imageViewModel.insert(array);
                });
    }
}

File: app/src/main/java/com/minerdev/greformanager/HouseDetailActivity.kt
package com.minerdev.greformanager

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.android.volley.Request
import com.google.gson.JsonObject
import com.minerdev.greformanager.Geocode.OnDataReceiveListener
import com.minerdev.greformanager.databinding.ActivityHouseDetailBinding
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.CameraUpdate
import com.naver.maps.map.MapFragment
import com.naver.maps.map.NaverMap
import com.naver.maps.map.overlay.Marker
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json

class HouseDetailActivity : AppCompatActivity() {
    lateinit var houseWrapper: HouseWrapper

    private val adapter = ImageSliderAdapter()
    private val viewModel: HouseModifyViewModel by viewModels()

    private lateinit var binding: ActivityHouseDetailBinding
    private var originalState = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_house_detail)
        binding.activity = this

        val intent = intent
        val data: House? = intent.getParcelableExtra("house_value")
        originalState = data!!.state.toInt()
        houseWrapper = HouseWrapper(data)

        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        viewModel.getOrderByPosition(data.id).observe(this, { images: List<Image> ->
            adapter.addImages(images)
            adapter.notifyDataSetChanged()
        })

        binding.viewPager2.adapter = adapter

        setMapFragment()

        val toggleButtonGroupManageFee = ToggleButtonGroup(this, "ManageFee")
        toggleButtonGroupManageFee.addToggleButtonsFromText(houseWrapper.manageFeeContains)
        for (toggleButton in toggleButtonGroupManageFee.toggleButtons) {
            binding.flowLayoutManageFee.addView(toggleButton)
        }

        val toggleButtonGroupOptions = ToggleButtonGroup(this, "Options")
        toggleButtonGroupOptions.addToggleButtonsFromText(houseWrapper.options)
        for (toggleButton in toggleButtonGroupOptions.toggleButtons) {
            binding.flowLayoutOptions.addView(toggleButton)
        }

        viewModel.loadImages(data.id)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val menuInflater = menuInflater
        menuInflater.inflate(R.menu.house_detail_toolbar, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()

            R.id.house_detail_menu_modify -> {
                val intent = Intent(this, HouseModifyActivity::class.java)
                intent.putExtra("mode", "modify")
                intent.putExtra("house_value", houseWrapper.data)
                startActivityForResult(intent, Constants.instance.HOUSE_MODIFY_ACTIVITY_REQUEST_CODE)
            }

            else -> {

            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun finish() {
        if (originalState != houseWrapper.data.state.toInt()) {
            val data = JsonObject()
            data.addProperty("state", houseWrapper.data.state)
            HttpConnection.instance.send(Request.Method.PATCH,
                    "houses/" + houseWrapper.data.id, data, object : HttpConnection.OnReceiveListener {
                override fun onReceive(receivedData: String) {
                    val house = Json.decodeFromString<House>(receivedData)
                    viewModel.modify(house)
                }
            })
        }

        super.finish()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == Constants.instance.HOUSE_MODIFY_ACTIVITY_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                data?.let {
                    val house: House? = data.getParcelableExtra("house_value")
                    val houseWrapper = HouseWrapper(house!!)
                    this.houseWrapper = houseWrapper
                    refreshUI()
                }
            }
        }
    }

    private fun refreshUI() {
        binding.invalidateAll()
        binding.flowLayoutManageFee.removeAllViews()
        binding.flowLayoutOptions.removeAllViews()

        val toggleButtonGroupManageFee = ToggleButtonGroup(this, "ManageFee")
        toggleButtonGroupManageFee.addToggleButtons(houseWrapper.manageFeeContains.split('|'))
        for (toggleButton in toggleButtonGroupManageFee.toggleButtons) {
            binding.flowLayoutManageFee.addView(toggleButton)
        }

        val toggleButtonGroupOptions = ToggleButtonGroup(this, "Options")
        toggleButtonGroupOptions.addToggleButtons(houseWrapper.options.split('|'))
        for (toggleButton in toggleButtonGroupOptions.toggleButtons) {
            binding.flowLayoutOptions.addView(toggleButton)
        }

        setMapFragment()
    }

    private fun setMapFragment() {
        var mapFragment = supportFragmentManager.findFragmentById(R.id.map) as MapFragment?

        if (mapFragment == null) {
            mapFragment = MapFragment.newInstance()
            supportFragmentManager.beginTransaction().add(R.id.map, mapFragment).commit()
        }

        mapFragment?.getMapAsync { naverMap: NaverMap ->
            naverMap.isLiteModeEnabled = true

            val uiSettings = naverMap.uiSettings
            uiSettings.isZoomControlEnabled = false
            uiSettings.setAllGesturesEnabled(false)

            Geocode.instance.getQueryResponseFromNaver(this@HouseDetailActivity, houseWrapper.address.substring(8))
            Geocode.instance.setOnDataReceiveListener(object : OnDataReceiveListener {
                override fun parseData(result: GeocodeResult?) {
                    result?.let {
                        val address = result.addresses!![0]
                        val latLng = LatLng(address.y.toDouble(), address.x.toDouble())
                        naverMap.moveCamera(CameraUpdate.scrollAndZoomTo(latLng, 16.0))

                        val marker = Marker(latLng)
                        marker.map = naverMap
                    }
                }
            })
        }
    }

    companion object {
        fun convertBooleanToVisibility(visible: Boolean): Int {
            return if (visible) View.VISIBLE else View.GONE
        }
    }
}